native_client/ctcdecode/setup.py

```diff
@@ -0,0 +1,147 @@
+                     include_dirs=None,
+                     debug=0,
+                     extra_preargs=None,
+                     extra_postargs=None,
+                     depends=None):
+    # those lines are copied from distutils.ccompiler.CCompiler directly
+    macros, objects, extra_postargs, pp_opts, build = self._setup_compile(
+        output_dir, macros, include_dirs, sources, depends, extra_postargs)
+    cc_args = self._get_cc_args(pp_opts, debug, extra_preargs)
+
+    # parallel code
+    def _single_compile(obj):
+        try:
+            src, ext = build[obj]
+        except KeyError:
+            return
+        self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
+
+    # convert to list, imap is evaluated on-demand
+    thread_pool = multiprocessing.pool.ThreadPool(args[0].num_processes)
+    list(thread_pool.imap(_single_compile, objects))
+    return objects
+
+
+def compile_test(header, library):
+    dummy_path = os.path.join(os.path.dirname(__file__), "dummy")
+    command = "bash -c \"g++ -include " + header \
+                + " -l" + library + " -x c++ - <<<'int main() {}' -o " \
+                + dummy_path + " >/dev/null 2>/dev/null && rm " \
+                + dummy_path + " 2>/dev/null\""
+    return os.system(command) == 0
+
+
+# hack compile to support parallel compiling
+distutils.ccompiler.CCompiler.compile = parallelCCompile
+
+FILES = glob.glob('../kenlm/util/*.cc') \
+        + glob.glob('../kenlm/lm/*.cc') \
+        + glob.glob('../kenlm/util/double-conversion/*.cc')
+
+FILES += glob.glob('third_party/openfst-1.6.7/src/lib/*.cc')
+
+FILES = [
+    fn for fn in FILES
+    if not (fn.endswith('main.cc') or fn.endswith('test.cc') or fn.endswith(
+        'unittest.cc'))
+]
+
+LIBS = ['stdc++']
+if platform.system() != 'Darwin':
+    LIBS.append('rt')
+
+ARGS = ['-O3', '-DNDEBUG', '-DKENLM_MAX_ORDER=6', '-std=c++11',
+        '-Wno-unused-local-typedef', '-Wno-sign-compare']
+
+if compile_test('zlib.h', 'z'):
+    ARGS.append('-DHAVE_ZLIB')
+    LIBS.append('z')
+
+if compile_test('bzlib.h', 'bz2'):
+    ARGS.append('-DHAVE_BZLIB')
+    LIBS.append('bz2')
+
+if compile_test('lzma.h', 'lzma'):
+    ARGS.append('-DHAVE_XZLIB')
+    LIBS.append('lzma')
+
+decoder_module = Extension(
+    name='ds_ctcdecoder._swigwrapper',
+    sources=['swigwrapper.i'] + FILES + glob.glob('*.cpp'),
+    swig_opts=['-c++', '-extranative'],
+    language='c++',
+    include_dirs=[
+        numpy_include,
+        '..',
+        '../kenlm',
+        'third_party/openfst-1.6.7/src/include',
+        'third_party/ThreadPool',
+    ],
+    libraries=LIBS,
+    extra_compile_args=ARGS
+)
+
+class BuildExtFirst(build):
+    sub_commands = [('build_ext', build.has_ext_modules),
+                    ('build_py', build.has_pure_modules),
+                    ('build_clib', build.has_c_libraries),
+                    ('build_scripts', build.has_scripts)]
+
+setup(
+    name='ds_ctcdecoder',
```

Comment:
Will we put that on PyPi ? If so, we might want to put proper links / docs as we do on other bindings

Reply:
Probably not? If we do and people assume you can train master with PyPi ds_ctcdecoder then we have to start being very careful about breaking the API "inside" of a DeepSpeech version.

Reply:
Ok, then, for training we will require people to build the wheel ? Or we distribute it in some way ?

Reply:
Yes, probably get it from our infra using `util/taskcluster.py`.